### ISSUE TITLE
Runbook page loops endlessly on runbook-not-found

### ISSUE THREAD

**Issue by matt-richardson** [OctopusDeploy/Issues#7173]
### Team

- [X] I've assigned a team label to this issue

### Severity

Low

### Version

2021.3.8275

### What happened?

I had a reminder pop up to delete a certain runbook after a period of time had elapsed (to clean things up).

Clicking on the link showed a flash of an error message, then took me to a page that looped endlessly. Checking the console, it showed a 404 on the project and then an uncaught exception:
```
Uncaught (in promise) n {StatusCode: 404, ErrorMessage: "Exception of type 'Octopus.Core.Model.Exceptions.EntityNotFoundException' was thrown.", Errors: Array(0)}
```

### Reproduction

https://deploy.octopus.app/app#/Spaces-1/projects/myproject/operations/runbooks/Runbooks-1/overview

### Error and Stacktrace

```shell
/app#/Spaces-1/proje…nbooks-1/overview:1 Uncaught (in promise) n {StatusCode: 404, ErrorMessage: "Exception of type 'Octopus.Core.Model.Exceptions.EntityNotFoundException' was thrown.", Errors: Array(0)}
XMLHttpRequest.send (async)
(anonymous) @ VM40:1
e.send @ index.esm.js:25
sendRequest @ ajax.ts:135
execute @ ajax.ts:99
(anonymous) @ client.ts:390
(anonymous) @ es6-promise.js:459
R @ es6-promise.js:871
dispatchRequest @ client.ts:377
get @ client.ts:168
get @ basicRepository.ts:63
(anonymous) @ ProjectContext.tsx:174
_ @ ProjectContext.tsx:105
(anonymous) @ ProjectContext.tsx:173
(anonymous) @ ProjectContext.tsx:232
al @ react-dom.production.min.js:211
ws @ react-dom.production.min.js:257
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vo @ react-dom.production.min.js:122
bs @ react-dom.production.min.js:257
rs @ react-dom.production.min.js:239
(anonymous) @ react-dom.production.min.js:123
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vo @ react-dom.production.min.js:122
$o @ react-dom.production.min.js:123
Xo @ react-dom.production.min.js:122
Zl @ react-dom.production.min.js:230
enqueueSetState @ react-dom.production.min.js:132
w.setState @ react.production.min.js:12
stepIntoSpace @ SpaceLoader.tsx:168
await in stepIntoSpace (async)
stepIntoSelectedSpaceId @ SpaceLoader.tsx:139
switchToSpace @ SpaceLoader.tsx:113
UNSAFE_componentWillReceiveProps @ SpaceLoader.tsx:79
Ci @ react-dom.production.min.js:135
Ba @ react-dom.production.min.js:179
wl @ react-dom.production.min.js:263
fs @ react-dom.production.min.js:246
ds @ react-dom.production.min.js:246
rs @ react-dom.production.min.js:239
(anonymous) @ react-dom.production.min.js:123
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vo @ react-dom.production.min.js:122
$o @ react-dom.production.min.js:123
Xo @ react-dom.production.min.js:122
os @ react-dom.production.min.js:240
notify @ Subscription.js:23
t.notifyNestedSubs @ Subscription.js:65
t.handleChangeWrapper @ Subscription.js:70
A @ redux.js:220
(anonymous) @ index.js:11
onUserAccessibleSpacesLoaded @ SpaceLoader.tsx:180
loadSpaces @ SpaceLoader.tsx:104
await in loadSpaces (async)
componentDidMount @ SpaceLoader.tsx:85
ll @ react-dom.production.min.js:212
gs @ react-dom.production.min.js:255
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vo @ react-dom.production.min.js:122
vs @ react-dom.production.min.js:248
rs @ react-dom.production.min.js:239
(anonymous) @ react-dom.production.min.js:123
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vo @ react-dom.production.min.js:122
$o @ react-dom.production.min.js:123
Xo @ react-dom.production.min.js:122
Zl @ react-dom.production.min.js:230
enqueueSetState @ react-dom.production.min.js:132
w.setState @ react.production.min.js:12
setState @ BaseComponent.tsx:31
authenticationSucceeded @ SignIn.tsx:297
await in authenticationSucceeded (async)
continueIfAlreadyAuthenticated @ SignIn.tsx:363
await in continueIfAlreadyAuthenticated (async)
connectClient @ SignIn.tsx:111
componentDidMount @ SignIn.tsx:90
<computed> @ BaseComponent.tsx:135
ll @ react-dom.production.min.js:212
gs @ react-dom.production.min.js:255
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vo @ react-dom.production.min.js:122
vs @ react-dom.production.min.js:248
rs @ react-dom.production.min.js:239
(anonymous) @ react-dom.production.min.js:123
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vo @ react-dom.production.min.js:122
$o @ react-dom.production.min.js:123
Xo @ react-dom.production.min.js:122
Zl @ react-dom.production.min.js:230
enqueueSetState @ react-dom.production.min.js:132
w.setState @ react.production.min.js:12
setState @ BaseComponent.tsx:31
connectClient @ ClientConnector.tsx:51
await in connectClient (async)
componentDidMount @ ClientConnector.tsx:40
<computed> @ BaseComponent.tsx:135
ll @ react-dom.production.min.js:212
gs @ react-dom.production.min.js:255
t.unstable_runWithPriority @ scheduler.production.min.js:19
Vo @ react-dom.production.min.js:122
vs @ react-dom.production.min.js:248
rs @ react-dom.production.min.js:239
Zl @ react-dom.production.min.js:230
Hs @ react-dom.production.min.js:281
(anonymous) @ react-dom.production.min.js:284
is @ react-dom.production.min.js:240
qs @ react-dom.production.min.js:284
t.render @ react-dom.production.min.js:290
64019 @ index.tsx:32
o @ bootstrap:19
(anonymous) @ main.12d804d1af89dd1d29a2.hashedasset.js:1
o.O @ chunk loaded:23
(anonymous) @ main.12d804d1af89dd1d29a2.hashedasset.js:1
r @ jsonp chunk loading:33
(anonymous) @ main.12d804d1af89dd1d29a2.hashedasset.js:1
```

**Comment by droyad** [comment]
This looks like a problem in the common loading control, redirecting

**Comment by TomPeters** [comment]
Normally pages have an isolated error context.

When we load the project as part of the project context, this particularly component happens to be rendered "outside" of any pages, which means it falls back to using the global error context provided by App.tsx

Meanwhile, other components (specifically the ProjectSwitcher) reuse this same error context.

The bug occurs because
- The project layout tries to load and receives an error. It places this on the global error context.
- An async action is triggered from the ProjectSwitcher (e.g. loading projects). As part of this action, the current error context is cleared
- This causes unexpected behaviour in the project layout. Specifically, [if there is no project and no errors](https://github.com/OctopusDeploy/OctopusDeploy/blob/9ee2c7927d98713626495a154934a949fbd25b38/newportal/app/areas/projects/components/ProjectLayout/ProjectLayout.tsx#L68), it assumes that it is currently loading and it displays a loading indicator.

**Comment by TomPeters** [comment]
Internal link: https://trello.com/c/vhmpZObL/158-projectlayout-uses-global-error-context-causing-unexpected-behaviour

**Comment by octoreleasebot** [comment]
Release Note: Fixed a UI issue where an infinite loading indicator could be displayed instead of a project loading error

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2021.3 | 2021.3.11006 |
| 2022.1+ | _all releases_ |

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2022.2 | 2022.2.0 |

**Comment by Octobob** [comment]
:tada: The fix for this issue has been released in:

| Release stream | Release     |
| -------------- | ----------- |
| 2021.3 | 2021.3.11006 |
| 2022.1+ | _all releases_ |
